### ISSUE TITLE
prefer_type_checking improperly corrects `as` coercions

### ISSUE THREAD

**Issue by jumhyn-browser** [realm/SwiftLint#5854]
### New Issue Checklist

- [x] I've Updated SwiftLint to the latest version.
- [x] I've searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues).

### Bug Description

This rule should not trigger on `as` coercions because these expressions supply important type context that can change typechecking behavior or allow it to succeed. If the below example is changed to `is` and `foo()` is a generic function which depended on return type context to determine the generic parameter, typechecking would fail. 

```swift
// This triggers a violation:
if let foo() as Bar? != nil {}
```

**Comment by SimplyDanny** [comment]
> ```swift
> // This triggers a violation:
> if let foo() as Bar? != nil {}
> ```

Is this even valid Swift? I wonder about the `foo()` being a function call. I guess, the `let` is wrong here?

Anyway, this example (without the `let`) doesn't trigger on `main` anymore. It got fixed by #5805 meanwhile.

@mildm8nnered: Do you agree? If so, please close this ticket.

**Comment by mildm8nnered** [comment]
> Is this even valid Swift? I wonder about the foo() being a function call. I guess, the let is wrong here?
> Anyway, this example (without the let) doesn't trigger on main anymore. It got fixed by #5805 meanwhile.

> @mildm8nnered: Do you agree? If so, please close this ticket.

Yep, it does not look valid to me, but in any case, this no longer triggers. Closing as resolved.

**Comment by jumhyn-browser** [comment]
Oop, yeah, @SimplyDanny the `let` is spurious here. Glad this has been fixed on main, thanks for taking a look!
